pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+				// Service requiring allocation
+				// TODO is the oldip relevant here?
+				return ic.allocate(service, change.key)
+			}
+		} else {
+			// A service for which allocation status exists
+			// but that is no longer of type LoadBalancer
+			// should be deallocated.
+			if len(service.Status.LoadBalancer.Ingress) > 0 {
+				// TODO is the oldip relevant here?
+				return ic.deallocate(service, change.key)
+			}
+		}
+	} else if oldIP != "" {
+		// Assume deletion if no service is present but an old ip exists.
+		key, err := controller.KeyFunc(change.oldService)
+		if err != nil {
+			// Recovery/retry not possible for this error
+			glog.Errorf("Couldn't get key for object %+v: %v", change.oldService, err)
+			return nil
+		}
+		ic.clearLocalAllocation(key, oldIP)
+	}
+
+	return nil
+}
+
+// clearLocalAllocation clears an in-memory allocation if it belongs
+// to the specified service key.
+func (ic *IngressIPController) clearLocalAllocation(key, ipString string) {
+	ip := net.ParseIP(ipString)
+	if ip == nil {
+		// An invalid ip address cannot be deallocated
+		glog.Errorf("Error parsing ip: %v", ipString)
+		return
+	}
+
+	ipKey, ok := ic.allocationMap[ipString]
+	if !ok {
+		// IP is not currently allocated
+		return
+	}
+	if key != ipKey {
+		// IP is not allocated to the service
+		return
+	}
+
+	// Remove allocation
+	err := ic.ipAllocator.Release(ip)
+	if err != nil {
+		// Release from contiguous allocator should never return an error.
+		glog.Errorf("Error releasing ip %v for service %v: %v", ipString, key, err)
+		return
+	}
+	delete(ic.allocationMap, ipString)
+}
+
+// recordAllocation updates local state with the ingress ip indicated
+// in a service's status and ensures that the ingress ip appears in
+// the service's list of external ips.
+func (ic *IngressIPController) recordAllocation(service *kapi.Service, key, oldIP string) error {
+	// If more than one ingress ip is present, it will be ignored
+	ipString := service.Status.LoadBalancer.Ingress[0].IP
+
+	// Clear a previously allocated ip
+	if oldIP != "" && ipString != oldIP {
+		ic.clearLocalAllocation(key, oldIP)
+	}
+
+	// Make a copy to modify to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	sCopy := t.(*kapi.Service)
+
+	ip := net.ParseIP(ipString)
+	if ip == nil {
+		// An invalid ingress ip is assigned to the service.
+		if !ic.syncProcessed {
+			// Fix the problem after processing the initial sync to
+			// avoid delaying cache updates any longer than necessary.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			return ic.clearPersistedAllocation(sCopy)
+		}
+	}
+
+	ipKey, ok := ic.allocationMap[ipString]
+	if ok && ipKey == key {
+		// Allocation is for this service.
+		if !ic.syncProcessed {
+			// Avoid ensuring an external ip until after the initial sync.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
```

Comment:
Style, no else blocks if the first block returns.  I would also invert this condition because the else is shorter.